### ISSUE TITLE
Text is not displayed

### ISSUE THREAD

**Issue by asher-stern** [webanno/webanno#420]
This bug refers to the attached document.
System version 2.3.1

The document's contents are not fully displayed by the system.
When changing the number of lines displayed in a page (to 40, for example), the system suddenly displays no-text at all.
Even when the number of lines is set to 5 or 10 sentences, the system shows only part of the text. The majority of the text is not displayed to the annotator.

Here is the full document text:

> Menu About Us 
> Products Switches ICS 
> Network 
> Brochures 
> News 
> Contact 
> 
> Search English Deutsch 
> English 
> 简化 
> 傳統 
> 日本の 
> 
> Legal Notes 
> Limitation of liability for internal content The content of our website has been compiled with meticulous care and to the best of our knowledge. However, we cannot assume any liability for the up-to-dateness, completeness or accuracy of any of the pages. Pursuant to section 7, para. 1 of the TMG (Telemediengesetz – Tele Media Act by German law), we as service providers are liable for our own content on these pages in accordance with general laws. However, pursuant to sections 8 to 10 of the TMG, we as service providers are not under obligation to monitor external information provided or stored on our website. Once we have become aware of a specific infringement of the law, we will immediately remove the content in question. Any liability concerning this matter can only be assumed from the point in time at which the infringement becomes known to us. Limitation of liability for external links Our website contains links to the websites of third parties (“external links”). As the content of these websites is not under our control, we cannot assume any liability for such external content. In all cases, the provider of information of the linked websites is liable for the content and accuracy of the information provided. At the point in time when the links were placed, no infringements of the law were recognisable to us. As soon as an infringement of the law becomes known to us, we will immediately remove the link in question. Copyright The content and works published on this website are governed by the copyright laws of Germany. Any duplication, processing, distribution or any form of utilisation beyond the scope of copyright law shall require the prior written consent of the author or authors in question. Data protection A visit to our website can result in the storage on our server of information about the access (date, time, page accessed). This does not represent any analysis of personal data (e.g., name, address or e-mail address). If personal data are collected, this only occurs – to the extent possible – with the prior consent of the user of the website. Any forwarding of the data to third parties without the express consent of the user shall not take place. We would like to expressly point out that the transmission of data via the Internet (e.g., by e-mail) can offer security vulnerabilities. It is therefore impossible to safeguard the data completely against access by third parties. We cannot assume any liability for damages arising as a result of such security vulnerabilities. The use by third parties of all published contact details for the purpose of advertising is expressly excluded. We reserve the right to take legal steps in the case of the unsolicited sending of advertising information; e.g., by means of spam mail. Source:  Mustervorlage.net Google Analytics: This website uses Google Analytics, a web analytics service provided by Google, Inc. (“Google”).  Google Analytics uses “cookies”, which are text files placed on your computer, to help the website analyze how users use the site. The information generated by the cookie about your use of the website (including your IP address) will be transmitted to and stored by Google on servers in the United States.  In case of activation of the IP anonymization, Google will truncate/anonymize the last octet of the IP address for Member States of the European Union as well as for other parties to the Agreement on the European Economic Area.  Only in exceptional cases, the full IP address is sent to and shortened by Google servers in the USA.  On behalf of the website provider Google will use this information for the purpose of evaluating your use of the website, compiling reports on website activity for website operators and providing other services relating to website activity and internet usage to the website provider.  Google will not associate your IP address with any other data held by Google.  You may refuse the use of cookies by selecting the appropriate settings on your browser. However, please note that if you do this, you may not be able to use the full functionality of this website.  Furthermore you can prevent Google’s collection and use of data (cookies and IP address) by downloading and installing the browser plug-in available under  https://tools.google.com/dlpage/gaoptout?hl=en-GB . You can refuse the use of Google Analytics by clicking on the following link. An opt-out cookie will be set on the computer, which prevents the future collection of your data when visiting this website: Disable Google Analytics Further information concerning the terms and conditions of use and data privacy can be found at  http://www.google.com/analytics/terms/gb.html  or at  https://www.google.de/intl/en_uk/policies/ .  Please note that on this website, Google Analytics code is supplemented by “gat._anonymizeIp();” to ensure an anonymized collection of IP addresses (so called IP-masking). To Top Home 
> About Us 
> Products 
> Network 
> Brochures 
> News 
> Contact 
> Legal Notes 
> Imprint 
> 
> ©2016 knitter-switch. All rights reserved. 
> knitter-switch, Neue Poststrasse 17, Postfach 10 02 33, 85598 Baldham, Germany 
> Tel +49 (0)8106-36 21 0 | Fax +49 (0)8106-36 21 40 Email: Click Here 
> 

A note:
Tokenization and sentence splitting was performed before importing this document to the system (as can be seen in the XMI file).
Segmentation was performed such that each line was annotated as a sentence. This was done in order to preserve the structure of the document, when presented to the annotator.
So it **might** be related to long-lines in the document, or to special characters, but it might be something else.

[google_legal__notes_www.knitter-switch.com_132.xmi.zip](https://github.com/webanno/webanno/files/721893/google_legal__notes_www.knitter-switch.com_132.xmi.zip)

**Comment by reckart** [comment]
Could you please try with WebAnno 3.0.0?

**Comment by asher-stern** [comment]
@reckart 

Hi.
Thanks.
About 3.0.0: I don't have 3.0.0 installed... so I can't try it.

**Comment by reckart** [comment]
@asher-stern can you privately send me a project export of the project where this happens such that I can simply import it into my local WebAnno to reproduce the problem?

**Comment by reckart** [comment]
The problem persists with 3.0.0.

It does not occur when displaying sentences individually, only when displaying at least two sentences - that hints at a problematic character located between two sentences.

The error is different depending on the browser being used but occurs in Chrome, Safari and Firefox alike.

The problematic location seems to be between sentence 13 and 14. I suspect the problem to be a unicode line break character (UTF `e2 80 a8`, or `\u2028`) that is located between the sentences.

```
a web analytics service provided by Google, Inc. (“Google”). \u2028Google Analytics uses “cookies”
```

When I try pasting this character into a `console.log("...")` command in a browsers console, it also causes trouble.

While I can imagine enhancing WebAnno to be more robust with respect to strange unicode characters, it seems to me to be the more practicable solution at the moment that you sanitize your texts before importing them into WebAnno. You have all kinds of very odd Unicode characters in there that I can well imagine may cause further problems when you make use of your annotated texts later.

**Comment by asher-stern** [comment]
So, is it a bug in the web-browsers?
As much as I can understand, this character is not "illegal" in some sense. It's just so rare that web-browsers do not handle it well.
Did I understand correctly?

**Comment by reckart** [comment]
The [JSON spec](https://tools.ietf.org/html/rfc7159) says that certain characters must be escaped in JSON. This includes the "normal" line-break character, but not this special Unicode variant. See Section "7.  Strings" in the specs referenced above.

I found a reference for the problem on [Stack Overflow](http://stackoverflow.com/questions/2965293/javascript-parse-error-on-u2028-unicode-character).

The JSON data is interpreted as JavaScript as part of the rendering process. While the data is ok in JSON, apparently, the JavaScript interpreters choke on it. The fix they suggest is to escape the characters on the server side. 

That's something I will try to add in WebAnno.

**Comment by reckart** [comment]
Ok, fix is committed. Would it be ok for you to switch to WebAnno 3 or do you strongly require a 2.3.x service release?

**Comment by asher-stern** [comment]
Hi.
Thanks very much!

So to use the fixed version I guess I have to pull the source code from GitHub, compile and install. Am I right?
If so, we will consider doing this. The question is how "dangerous" it is to use a snapshot of the code, which is not an official release.

**Comment by reckart** [comment]
The question may be more how "dangerous" it is to upgrade from WebAnno 2 to 3, because the lastest code is obviously the v3 code. I would recommend you set up a separate instance of WebAnno 3, export your projects from your WebAnno 2, import them into the new v3 instance and give it a try.

I have just made the first bug-fix release [WebAnno 3.0.1](https://github.com/webanno/webanno/releases/tag/webanno-3.0.1).

**Comment by asher-stern** [comment]
Great. Thanks.
